### PR TITLE
Migrate AWS settings to new settings infrastructure

### LINKED ISSUE #16293: Migrate AWS settings to new settings infrastructure
For EC2 discovery and S3 repository

### PR DESCRIPTION
**Pull request by dadoonet**
Reintroducing commit fb7723c but now deals with setting names conflicts
Also adds java documentation for each setting

Closes #16293.
Related to https://github.com/elastic/elasticsearch/pull/16477#discussion_r52469084

### CODE REVIEW

**Comment by dadoonet** [comment]
@danielmitterdorfer I'm assigning it to you as you already reviewed the first PR. Sorry! :p

File: plugins/discovery-ec2/src/main/java/org/elasticsearch/plugin/discovery/ec2/Ec2DiscoveryPlugin.java [hunk 1]
```diff
@@ -104,12 +105,51 @@ public void onModule(DiscoveryModule discoveryModule) {
...
-        settingsModule.registerSettingsFilterIfMissing(AwsEc2Service.CLOUD_AWS.PROXY_PASSWORD);
-        settingsModule.registerSettingsFilterIfMissing(AwsEc2Service.CLOUD_EC2.KEY);
-        settingsModule.registerSettingsFilterIfMissing(AwsEc2Service.CLOUD_EC2.SECRET);
-        settingsModule.registerSettingsFilterIfMissing(AwsEc2Service.CLOUD_EC2.PROXY_PASSWORD);
+        settingsModule.registerSettingsFilterIfMissing(AwsEc2Service.KEY_SETTING.getKey());
+        settingsModule.registerSettingsFilterIfMissing(AwsEc2Service.SECRET_SETTING.getKey());
+        settingsModule.registerSettingsFilterIfMissing(AwsEc2Service.PROXY_PASSWORD_SETTING.getKey());
+        settingsModule.registerSettingsFilterIfMissing(AwsEc2Service.CLOUD_EC2.KEY_SETTING.getKey());
+        settingsModule.registerSettingsFilterIfMissing(AwsEc2Service.CLOUD_EC2.SECRET_SETTING.getKey());
+        settingsModule.registerSettingsFilterIfMissing(AwsEc2Service.CLOUD_EC2.PROXY_PASSWORD_SETTING.getKey());
```
Comment (danielmitterdorfer) [inline plugins/discovery-ec2/src/main/java/org/elasticsearch/plugin/discovery/ec2/Ec2DiscoveryPlugin.java]: Is this by intention that a few global settings are not filtered (see invocations of `#registerSettingIfMissing()`)?

File: plugins/repository-s3/src/main/java/org/elasticsearch/plugin/repository/s3/S3RepositoryPlugin.java [hunk 2]
```diff
@@ -88,14 +89,79 @@ public void onModule(RepositoriesModule repositoriesModule) {
...
+
+        // Filter global settings
+        settingsModule.registerSettingsFilterIfMissing(AwsS3Service.KEY_SETTING.getKey());
+        settingsModule.registerSettingsFilterIfMissing(AwsS3Service.SECRET_SETTING.getKey());
+        settingsModule.registerSettingsFilterIfMissing(AwsS3Service.PROXY_PASSWORD_SETTING.getKey());
+        settingsModule.registerSettingsFilterIfMissing(AwsS3Service.CLOUD_S3.KEY_SETTING.getKey());
+        settingsModule.registerSettingsFilterIfMissing(AwsS3Service.CLOUD_S3.SECRET_SETTING.getKey());
+        settingsModule.registerSettingsFilterIfMissing(AwsS3Service.CLOUD_S3.PROXY_PASSWORD_SETTING.getKey());
+        settingsModule.registerSettingsFilter(S3Repository.Repository.KEY_SETTING.getKey());
+        settingsModule.registerSettingsFilter(S3Repository.Repository.SECRET_SETTING.getKey());
```
Comment (danielmitterdorfer) [inline plugins/repository-s3/src/main/java/org/elasticsearch/plugin/repository/s3/S3RepositoryPlugin.java]: Is this by intention that a few global settings are not filtered (see invocations of #registerSettingIfMissing())?

**Comment by danielmitterdorfer** [comment]
@dadoonet: I assumed in my review that the only thing that has changed between here and #16477 is the handling of duplicate settings. I've left two minor comments, otherwise LGTM.

File: plugins/repository-s3/src/main/java/org/elasticsearch/plugin/repository/s3/S3RepositoryPlugin.java (hunk 2 above)
Comment (dadoonet) [inline plugins/repository-s3/src/main/java/org/elasticsearch/plugin/repository/s3/S3RepositoryPlugin.java]: Yes. Filtering is only made to hide some settings (like passwords/tokens) when you are calling any API which could print settings.
Here, we only want to filter key/secret and proxy password.

Make sense?

Comment (danielmitterdorfer) [inline plugins/repository-s3/src/main/java/org/elasticsearch/plugin/repository/s3/S3RepositoryPlugin.java]: Thanks for the explanation. Makes sense now, yes. And I learned something too.

File: plugins/discovery-ec2/src/main/java/org/elasticsearch/cloud/aws/AwsEc2ServiceImpl.java [hunk 3]
```diff
@@ -138,27 +116,26 @@ public long delayBeforeNextRetry(AmazonWebServiceRequest originalRequest,
...
+                    new StaticCredentialsProvider(new BasicAWSCredentials(key, secret))
             );
         }
 
         this.client = new AmazonEC2Client(credentials, clientConfiguration);
 
-        if (settings.get(CLOUD_EC2.ENDPOINT) != null) {
-            String endpoint = settings.get(CLOUD_EC2.ENDPOINT);
+        String endpoint = CLOUD_EC2.ENDPOINT_SETTING.get(settings);
+        if (endpoint != null) {
```
Comment (ywelsch) [inline plugins/discovery-ec2/src/main/java/org/elasticsearch/cloud/aws/AwsEc2ServiceImpl.java]: this always yields true

File: plugins/discovery-ec2/src/main/java/org/elasticsearch/cloud/aws/AwsEc2ServiceImpl.java [hunk 4]
```diff
@@ -74,30 +72,15 @@ public synchronized AmazonEC2 client() {
...
-        String account = settings.get(CLOUD_EC2.KEY, settings.get(CLOUD_AWS.KEY));
-        String key = settings.get(CLOUD_EC2.SECRET, settings.get(CLOUD_AWS.SECRET));
+        clientConfiguration.setProtocol(CLOUD_EC2.PROTOCOL_SETTING.get(settings));
+        String key = CLOUD_EC2.KEY_SETTING.get(settings);
+        String secret = CLOUD_EC2.SECRET_SETTING.get(settings);
 
-        String proxyHost = settings.get(CLOUD_AWS.PROXY_HOST);
-        proxyHost = settings.get(CLOUD_EC2.PROXY_HOST, proxyHost);
+        String proxyHost = CLOUD_EC2.PROXY_HOST_SETTING.get(settings);
         if (proxyHost != null) {
```
Comment (ywelsch) [inline plugins/discovery-ec2/src/main/java/org/elasticsearch/cloud/aws/AwsEc2ServiceImpl.java]: yields always true

Comment (dadoonet) [inline plugins/discovery-ec2/src/main/java/org/elasticsearch/cloud/aws/AwsEc2ServiceImpl.java]: Indeed!

File: plugins/discovery-ec2/src/main/java/org/elasticsearch/cloud/aws/AwsEc2ServiceImpl.java (hunk 3 above)
Comment (dadoonet) [inline plugins/discovery-ec2/src/main/java/org/elasticsearch/cloud/aws/AwsEc2ServiceImpl.java]: Indeed!

**Pull request merged**